### ISSUE TITLE
Missing model

### ISSUE THREAD

**Issue by ocefpaf** [ocefpaf/sscsw#2]
@rsignell-usgs mentioned we should be picking this [1] model up.

[1] http://ona.coas.oregonstate.edu:8080/thredds/dodsC/NANOOS/OCOS.html

**Comment by rsignell-usgs** [comment]
Yes, the problem here is that although it shows up in NGDC geoportal:
![2015-12-18_15-59-31](https://cloud.githubusercontent.com/assets/1872600/11907282/a04032d8-a5a0-11e5-8619-d1638acac2a8.jpg)
the [metadata](http://ona.coas.oregonstate.edu:8080/thredds/iso/NANOOS/OCOS?catalog=http%3A%2F%2Fona.coas.oregonstate.edu%3A8080%2Fthredds%2Fcatalog.html&dataset=OCOS) has no start/stop time.

And this is because of a bug that existed in old versions of ncISO that assumed that the time coordinate variable was called `time`, while in ROMS runs, the time coordinate variable is called `ocean_time`.

**Comment by rsignell-usgs** [comment]
To fix this, we need OSU to modify their NcML and upgrade THREDDS.
- OSU folks need to modify their NcML and add `standard_name='time'` to the `ocean_time` variable:

```
<variable name="ocean_time">
      <attribute name="standard_name" value="time">
</variable>
```
- OSU needs to upgrade their THREDDS Data Server.     The best way to do this would be to use Docker, now that THREDDS is in a Docker container. 

Docker is cool because the docker container contains java, tomcat, thredds, everything you need.   And when a new version of thredds comes along that requires a new java or tomcat, you don't have to go through all that hassle.  You can just update your docker container.
1.  Install docker on the OSU server.   See instructions at docker.com.
   For example, ubuntu instructions are here:
   https://docs.docker.com/engine/installation/ubuntulinux/
2. Download the THREDDS docker container
   docker pull axiom/docker-thredds
3. Stop your existing tomcat that runs thredds
4. Run the thredds docker container using the instructions here: https://hub.docker.com/r/axiom/docker-thredds/

Here's an example of what the docker run command might look like:
- thredds catalogs are in `/usr/local/tomcat-thredds/content/thredds`
- The thredds catalogs refer to netcdf files on subdirectories under `/data2`
- the Tomcat running thredds is running on port 8080 and 8443

```
docker run -d -p 8080:8080 -p 8443:8443  -v /data2:/data2:ro  -v
/usr/local/tomcat-thredds/content/thredds:/opt/tomcat/content/thredds --name
thredds axiom/docker-thredds
```

**Comment by rsignell-usgs** [comment]
@crisien or @emiliom, would you be able to do this?  
We were hoping to have this working before the OOI meeting in January.

**Comment by emiliom** [comment]
Thanks for the detailed instructions. We definitely would like to have this
working for the ooi meeting, and yes, Craig is still the point person.
Craig, let's chat Monday or Tuesday.

Greetings on my way back from AGU.

**Comment by emiliom** [comment]
@ocefpaf, just a note that we (meaning mostly @crisien and @rsignell-usgs!) have made a lot of progress on this in the last few days. It looks like everything is set on the NANOOS THREDDS / Craig's end, but we're waiting (and crossing fingers) for propagation to the NGDC CSW. To help track that propagation:
- Here's the [live iso metadata directly from Craig's THREDDS](http://ona.coas.oregonstate.edu:8080/thredds/iso/NANOOS/OCOS?catalog=http%3A%2F%2Fona.coas.oregonstate.edu%3A8080%2Fthredds%2Fcatalog.html&dataset=OCOS)
- And here's the [corresponding iso from the NGDC Geoportal](http://www.ngdc.noaa.gov/geoportal/rest/document?id={64BCA266-3EF5-4C92-83C8-05BAE244BD78})

Once the NGDC iso has a `gml:beginPosition` element and a `sea_water_temperature` keyword substring, we're good to go. I hope.

**Comment by ocefpaf** [comment]
Thanks for letting me know @emiliom. And thanks @rsignell-usgs and @crisien for fixing this.  I will try the notebook again in 2 days to check this.

**Comment by emiliom** [comment]
Just for tracking: The iso metadata on the NGDC Geoportal now has the `sea_water_temperature` (which is unsurprising, since that's a simple update). But the time start and end Position elements are not there yet, and probably won't be [until NGDC's stand-alone ncISO is updated.](https://github.com/ioos/registry/issues/88)

@ocefpaf, if you still have some time this coming week to work on the OOI notebook that uncovered this issue, one possibility is to acknowledge in the notebook that the CSW query doesn't return the OSU ROMS record yet (but will in the "near" future), then proceed with a hard-wired direct ingest from the OSU ROMS THREDDS service. Like the CO-OPS time series ingest, but better ;)

**Comment by ocefpaf** [comment]
> @ocefpaf, if you still have some time this coming week to work on the OOI notebook that uncovered this issue, one possibility is to acknowledge in the notebook that the CSW query doesn't return the OSU ROMS record yet (but will in the "near" future), then proceed with a hard-wired direct ingest from the OSU ROMS THREDDS service. Like the CO-OPS time series ingest, but better ;)

Thanks @emiliom.  I still cannot find that model. (Actually more data disappeared!  See [here](http://nbviewer.ipython.org/urls/gist.githubusercontent.com/ocefpaf/07a27754a69bd3c22004/raw/0a990b66c1069b16be71cdc77a73524cc0ac8524/csw_demo.ipynb).)

I will leave that decision to @kbailey-noaa.  She will be presenting this at the OOI meeting.

**Comment by emiliom** [comment]
>  I still cannot find that model.

Yeah, it's not findable with the date filter, b/c of the issue being addressed. However:

> Actually more data disappeared!

Yes, and yikes. NDBC is now gone, and the OSU ROMS model is no longer findable even with a simple query (on "NANOOS") that did return it just a week ago. I'll try to track this a bit today -- see #1

**Comment by ocefpaf** [comment]
Thanks for looking into this @emiliom!  It seems that we are finding more problems than we expected with this exercise.

**Comment by emiliom** [comment]
> It seems that we are finding more problems than we expected with this exercise.

Well, we're finding **new**, unexpected problems (NDBC), resurrecting old ones that had been shoved under the carpet (CO-OPS), and fixing some issues (OSU ROMS THREDDS). But overall the exercise has reached a no-fun stage :(

**Comment by kbailey-noaa** [comment]
Since there are various problems, I will stick with the Boston Light example only (rather than include a 2nd example) but I will acknowledge that there is a very similar notebook we are working on that will also be able to include the OSU ROMS model.  I'll be pointing the group to the github pages and notebooks for reference, so this will be another good addition for users to tweak (once all the kinks are worked out).  So please don't rush this for the OOI meeting but I do think it would be worthwhile to provide this notebook to users.  Thanks very much for the efforts!

**Comment by emiliom** [comment]
@kbailey-noaa, for what it's worth, be aware that the NDBC and COOPS problems we've been discussing here apply equally to any other region. Though NDBC is now fixed (an NGDC registry/GeoPortal update synchronization timing issue).

The OSU ROMS THREDDS web service issues were fixed on the NANOOS end last week, but the NGDC ncISO harvester doesn't handle it yet b/c it needs to be updated to a version of ncISO that fixed an important issue.

**IF** @ocefpaf was already working on showing the OSU ROMS model on this notebook, we may actually be close to done. But he's already put a lot of time into this, and I know he was in the middle of an overhaul that had to be stopped due to the NGDC CSW hiccups, so I realize that @ocefpaf probably can't spend more time (and it's Dec. 30th!).

I totally understand your hesitation to use (or rather, wait for) this notebook. But it's also kind of disappointing for me (ie, NANOOS) given all the effort I've put into it over the last couple of weeks ... 

Happy New Year!

**Comment by rsignell-usgs** [comment]
@emiliom, the start/stop times could also appear if NANOOS started using the preferred method of maintaining their own ISO metadata in a WAF that NGDC (or someone else) can just harvest.

**Comment by emiliom** [comment]
@rsignell-usgs: right. We'll work on that starting next week, when Craig is back. But that will also involve a change to the registration of the OSU ROMS endpoint, then 1-day propagation time at NGDC. So, not doable for the OOI meeting, but definitely a target for NANOOS for the next couple of weeks. Thanks for the reminder.

**Comment by ocefpaf** [comment]
As we say in Brazil this time of the year: "there is still plenty of time until next year."

Here is an as-is version of the notebook that, thanks to @amilan17, now finds NDBC data:

http://nbviewer.ipython.org/github/ocefpaf/sscsw/blob/master/IOOS_data_demo.ipynb

I still need to apply some of the comments from both @emiliom and @rsignell-usgs, clean/update the text, and steal some ideas from @emiliom's notebooks.  However, this version is almost ready and it does not use any obscure custom library (`utilities`)!

PS0: @emiliom note that I am displaying the models as grid outlines (GeoJSONs with the `nc.summary` as their popup) rather than the raster grid image as before.  I believe this is a better way to display "model assets."

PS1: The absence of this model was never a crucial part of the notebook.  I added it here just to keep track of things.

PS2: I am preparing a binder for this repo soon.

**Comment by emiliom** [comment]
And @ocefpaf saves the day, yet again :) Thanks!

I can see you've already implemented a lot of improvements on this notebook. Neat!!

> PS0: @emiliom note that I am displaying the models as grid outlines (GeoJSONs with the nc.summary as their popup) rather than the raster grid image as before. I believe this is a better way to display "model assets."

Very nice. Yes, using the model domain outline makes the asset very visible, unlike the light gray mesh used before (though the mesh is really nice once you zoom in to it). I also like the generic approach you're using to step through the models. And I also like how you added the abstract to a pop up!

> PS1: The absence of this model was never a crucial part of the notebook. I added it here just to keep track of things.

The second model seen on the notebook (California ROMS) is not the OSU ROMS. It's great that it's there, of course, since it's returned by the CSW query and it's geographically relevant (though barely). For now, an easy but hard-wired way to add the OSU ROMS model would be, in Cell 16:

``` python
# Change this:
for url in dap_urls:
# to this:
for url in dap_urls + ['http://ona.coas.oregonstate.edu:8080/thredds/dodsC/NANOOS/OCOS.html']:
```

**Comment by kbailey-noaa** [comment]
Hi - The notebook looks great so I'll use some pieces for a slide.  I think it would be really useful.  (I thought there were a lot more issues - my mistake).  I plan to include the link to the notebook as well.  I assume the bulleted notes at the top would be deleted?  Thanks!

**Comment by ocefpaf** [comment]
> though the mesh is really nice once you zoom in to it

`gridgeo` has a **geo_interface** for the mesh too, but that is too expensive to show in the browser.  I need to find a better way to rasterize it to show meshes for efficiently. (See cell [5] from [this example](http://nbviewer.ipython.org/github/pyoceans/gridgeo/blob/master/notebooks/example_gridgeo_ugrid.ipynb).)

>  For now, an easy but hard-wired way to add the OSU ROMS model would be, in Cell 16:

Since the goal is to show the automated tools I am not hard-wiring anything.  (Unless @kbailey-noaa wants to show what we are missing/what else could be found.)

> I thought there were a lot more issues - my mistake

There were!  Thanks to @emiliom's help we are solving them one-by-one.

**Comment by emiliom** [comment]
> Since the goal is to show the automated tools I am not hard-wiring anything.

That's reasonable, and makes for a very consistent demo. It will keep COOPS data out of the notebook as well, but that's life.

**Comment by emiliom** [comment]
@ocefpaf, one suggestion for the time series plot in Cell 12: It'd be great to tweak the datetime x-axis tick labels to something more informative than the default of hours without dates. That pandas dataframe default for a time interval of a few days is annoying ...

**Comment by ocefpaf** [comment]
>  for the time series plot in Cell 12: It'd be great to tweak the datetime x-axis tick labels to something more informative than the default of hours without dates. 

Done in https://github.com/ocefpaf/sscsw/commit/d9ab7fe6f8f7d3a9cf5b8b30cd24634e1c364261

**Comment by emiliom** [comment]
Great. And nice work on the additional polish you've done on the notebook.
